backend/redis/utils.go

```diff
@@ -344,6 +345,89 @@ func (r *Client) GetResources(ctx context.Context, s *model.Session, resources m
 	return allResources, nil
 }
 
+// Adds a session to be processed `delaySeconds` in the future
+func (r *Client) AddSessionToProcess(ctx context.Context, sessionId int, delaySeconds int) error {
+	score := float64(time.Now().Unix() + int64(delaySeconds))
+
+	cmd := r.Client.ZAdd(ctx, CacheKeySessionsToProcess, redis.Z{
+		Score:  score,
+		Member: sessionId,
+	})
+	if err := cmd.Err(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Removes a session after processing (successfully or with errors)
+// Only removes the session if its processing time is from a 'lock',
+// in case more events were added after it started processing.
+func (r *Client) RemoveSessionToProcess(ctx context.Context, sessionId int) error {
+	var script = redis.NewScript(`
+		local key = KEYS[1]
+		local sessionId = ARGV[1]
+		
+		local score = tonumber(redis.call("ZSCORE", key, sessionId))
+		if score ~= nil and score ~= math.floor(score) then
+			return redis.call("ZREM", key, sessionId)
+		end
+		
+		return 0
+	`)
+
+	keys := []string{CacheKeySessionsToProcess}
+	values := []interface{}{sessionId}
+	cmd := script.Run(ctx, r.Client, keys, values...)
+
+	if err := cmd.Err(); err != nil && !errors.Is(err, redis.Nil) {
+		return err
+	}
+
+	return nil
+}
+
+// Retrieves up to `limit` sessions to process. Sets the processing time
+// for each to `lockPeriod` minutes after the current time, so they
+// can be retried in case processing fails.
+func (r *Client) GetSessionsToProcess(ctx context.Context, lockPeriod int, limit int) ([]int64, error) {
+	now := time.Now().Unix()
+	timeAfterLock := float64(now+int64(60*lockPeriod)) + .5
+	var script = redis.NewScript(`
+		local key = KEYS[1]
+		local now = ARGV[1]
+		local timeAfterLock = ARGV[2]
+		local limit = ARGV[3]
+		local range = redis.call("ZRANGEBYSCORE", key, 0, now, "LIMIT", 0, limit)
+
+		local input = {}
+		local count = 0
+		for k, item in pairs(range) do 
+			table.insert(input, timeAfterLock)
+			table.insert(input, item)
+			count = count + 1
+		end
+
+		if count == 0 then
+			return {}
+		end
+
+		redis.call("ZADD", key, unpack(input))
```

Comment:
what's the point of reinserting the values? does `zrangebyscore` pop off things that it returns?

Reply:
the values were not removed, `ZADD` is just updating their scores here (with the new time)